### PR TITLE
build: Enable building for win-arm and win-arm64

### LINKED ISSUE #5286: Pwrshplugin Config File Paths Should be Validated Before Use
The paths to CoreCLR and PSHOME read by pwrshplugin should be validated before use to protect against bad input.

A potential API to use is [PathIsDirectory](https://msdn.microsoft.com/en-us/library/windows/desktop/bb773621(v=vs.85).aspx).

It is part of api-ms-win-shlwapi-ie-l1-1-0.dll, but that API set is not present on Windows 7 (shlwapi.dll is though).

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
Requires VS2017 latest preview to build for arm64.  Validated win-arm build on Win10 IOT with remoting, but current install script doesn't work correctly on Win10 IOT, need to update.

Also fixed issue with pwrshplugin.dll not validating path before using them.

Fix https://github.com/PowerShell/PowerShell/issues/5286

### CODE REVIEW

**Comment by SteveL-MSFT** [comment]
Waiting on mmi.dll to be published for win-arm and win-arm64

**Comment by SteveL-MSFT** [comment]
Confirmed that basic PSCore6 works on win-arm, I think we can merge this.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by SteveL-MSFT** [comment]
@adityapatwardhan hold off on merging this, looks like I need to build pwshplugin.dll for arm

File: build.psm1
Comment (daxian-dbw) [inline build.psm1]: > !$env:PROCESSOR_ARCHITECTURE -match "arm"

`!$env:PROCESSOR_ARCHITECTURE` will be evaluated first in this case.

**Comment by daxian-dbw** [comment]
Once the changes to building `pwshplugin.dll` is done, the `psrp.windows` package needs to be updated, and make sure the arm version plugin dll is signed before creating a new nuget package.

File: build.psm1
Comment (SteveL-MSFT) [reply build.psm1]: Will fix

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT please update here when `psrp.windows` package is updated.

**Comment by daxian-dbw** [comment]
Note that PR #5617 would also require `psrp.windows` nuget package to be refreshed, so it's better to coordinate them to avoid duplicate work.

**Comment by SteveL-MSFT** [comment]
Have some paths hardcoded to validate build.  Will update, but requires user to have VS2017 latest installed to build for win-arm.

**Comment by SteveL-MSFT** [comment]
Verified pwrshplugin.dll works on both arm and arm64.  @adityapatwardhan can you re-review this PR as I've made substantial changes for the native compilation?

File: src/powershell-native/Install-PowerShellRemoting.ps1
Comment (adityapatwardhan) [inline src/powershell-native/Install-PowerShellRemoting.ps1]: `New-Item $regKey -Force  > $null `

**Review by adityapatwardhan — APPROVED** [review]
LGTM with minor comment.

**Comment by adityapatwardhan** [comment]
@daxian-dbw Can you have a look too?

File: src/powershell-native/Install-PowerShellRemoting.ps1
Comment (SteveL-MSFT) [reply src/powershell-native/Install-PowerShellRemoting.ps1]: Will change

**Comment by daxian-dbw** [comment]
@adityapatwardhan Yes, will look tonight or first thing tomorrow morning.

File: build.psm1
Comment (daxian-dbw) [inline build.psm1]: Question: Is it going to be required to use VS2017 to build the plugin dll?

Comment (adityapatwardhan) [reply build.psm1]: @daxian-dbw The build worker image we use in AppVeyor is `Visual Studio 2015`. Do we need to update that?

Comment (adityapatwardhan) [reply build.psm1]: Actually we are using `Visual Studio 2017` through the AppVeyor.yml

**Review by daxian-dbw — APPROVED** [review]
LGTM. Left a question comment.

File: build.psm1
Comment (daxian-dbw) [reply build.psm1]: We don't build the native binaries in CI, so it shouldn't matter. Maybe we should switch to an image that doesn't have VS pre-installed.

File: build.psm1
Comment (TravisEz13) [inline build.psm1]: `x64_arm` and `x64_arm64` seems weird to me.  Why are we combining two architectures?

**Review by TravisEz13 — APPROVED** [review]

File: build.psm1
Comment (SteveL-MSFT) [reply build.psm1]: I'll add a comment.  This is the form that the Visual Studio build environment expects for cross compilation.  The first is the host and the second the target.  VS supports x86_arm64, for example which I'm not supporting here.

File: build.psm1
Comment (SteveL-MSFT) [reply build.psm1]: @daxian-dbw VS2017 is required to build arm64

File: build.psm1
Comment (adityapatwardhan) [reply build.psm1]: @SteveL-MSFT Can you comment?

**Comment by SteveL-MSFT** [comment]
@TravisEz13 @daxian-dbw @adityapatwardhan please review the change to the pwrshplugin config reader to address https://github.com/PowerShell/PowerShell/issues/5286

Decided to do that as part of this PR since I'm publishing a new drop to myget might as well do it once rather than twice.

**Comment by SteveL-MSFT** [comment]
Published https://powershell.myget.org/feed/powershell-core/package/nuget/psrp.windows/6.0.0-rc

File: src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp
Comment (daxian-dbw) [inline src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp]: It seems the `size() > 0` check should be moved to before the `CreateFileW` call.
Or maybe the `size() > 0` is not needed, because if `getValueFromLine` returns an empty string, I thinkg `CreateFileW` will fail.

File: src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp
Comment (daxian-dbw) [inline src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp]: Why calling `this->getValueFromLine(line, coreClrDirTag);` again instead of use `coreClrDir`?

File: src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp
Comment (daxian-dbw) [inline src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp]: Same comment about the `size()` check here.

**Review by daxian-dbw — COMMENTED** [review]
A few **non-blocking** comments.

File: src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp
Comment (SteveL-MSFT) [reply src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp]: Mistake, should have been `coreClrDir`.  Will fix.

**Review by mirichmo — APPROVED** [review]

File: src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp
Comment (SteveL-MSFT) [reply src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp]: Wrote a small program to test this and an empty string does return an invalid handle.  I can remove the size check.

File: src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp
Comment (SteveL-MSFT) [reply src/powershell-native/nativemsh/pwrshcommon/ConfigFileReader.cpp]: same as above

**Comment by daxian-dbw** [comment]
@SteveL-MSFT Making those changes means we need to create a new `psrp.windows` package. But that work doesn't block merging this PR.

**Comment by SteveL-MSFT** [comment]
@daxian-dbw yeah, I realize that.  After this is merged, I'll create one called RC2.

**Review by daxian-dbw — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@daxian-dbw @TravisEz13 Any ideas why WIP might be blocking this?

**Comment by TravisEz13** [comment]
it hasn't run, edit the title and it will run.

**Pull request merged**
